Support for ServiceAccount name matching

tgraf opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4460)
<!-- Reviewable:end -->

eloycoto commented on test/k8sT/Policies.go:
```diff
@@ -282,6 +284,49 @@ var _ = Describe("K8sValidatedPolicyTest", func() {
 				helpers.CurlFail("http://%s/public", clusterIP))
 			res.ExpectSuccess("%q cannot curl to %q public", appPods[helpers.App2], clusterIP)
+
+			By("ServiceAccount")
```
I know that is sitll WIP, but please do the test in another `By()` ;-)

tgraf replied:
I don't understand the comment. Change the text in the By()? Move it to somewhere else?

tgraf replied:
Fixed

tgraf commented on test/k8sT/Policies.go:
```diff
@@ -284,6 +286,49 @@ var _ = Describe("K8sValidatedPolicyTest", func() {
 			res.ExpectSuccess("%q cannot curl to %q public", appPods[helpers.App2], clusterIP)
 		}, 500)
 
+		It("ServiceAccuont Based Enforcement", func() {
```
Typo

tgraf replied:
Fixed

rlenglet reviewed (approved):
LGTM.
However, this shouldn't be targeted at the 1.1 milestone, since it is out of scope per https://github.com/cilium/cilium/issues/3585.

tgraf committed:
test: ServiceAccount integration tests

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Add documentation and example for service account matching

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Represent ServiceAccountName as endpoint label

An endpoint representing a Kubernetes pod will automatically have its
ServiceAccountName represented as endpoint label. The endpoint label can be
used in policies to select or grant access to service accounts.

Signed-off-by: Thomas Graf <thomas@cilium.io>

rlenglet commented on pkg/workloads/kubernetes.go:
```diff
@@ -62,5 +62,10 @@ func fetchK8sLabels(containerLbls map[string]string) (map[string]string, error)
+
+	if result.Spec.ServiceAccountName != "" {
+		k8sLabels[k8sConst.PolicyLabelServiceAccount] = result.Spec.ServiceAccountName
+	}
```
Prevent users from spoofing a service account by defining this label in specs:
```
} else {
    delete k8sLabels[k8sConst.PolicyLabelServiceAccount]
}
```

tgraf replied:
Good idea I think but it doesn't prevent spoofing as the service account name can be freely specified in the pod spec so I'm not sure what security this increases. There is also "default" service account associated if nothing is specified so it will likely be dead code but probably the right thing to be sure. I think we will need an admission controller to make both service account and labels reliable.

rlenglet replied:
Maybe. E.g. the Istio docs recommend to use Kubernetes's recommended admission controllers set, which includes `ServiceAccount`. But that admission controller doesn't control the usage of service accounts. And I've not found anything that can restrict the usage of service accounts. AFAICT, if a user has the right to create workloads in a namespace, she can use any service account that exists in that namespace.
Let's leave it as is...

tgraf replied:
The corresponding admission controller does not exist yet but it is fairly trivial to write. It already exists for other privileges such as resources, privileged execution, etc.
